azkaban-common/src/main/java/azkaban/executor/ExecutionControllerUtils.java

```diff
@@ -168,26 +168,46 @@ public static void finalizeFlow(final EventHandler eventHandler,
   }
 
   /**
-   * This method tries to restart the flow for certain statuses otherwise simply return. There are
-   * three scenarios that this method is called: 1. a flow is cleaned up by ContainerCleanupManager;
-   * 2. a flow has dispatch failure; 3. a flow encounters pod failure Each flow execution can be
-   * retried once. If the original status is EXECUTION_STOPPED, then it will be retried only if
-   * allow.restart.on.execution.stopped is set to true
+   * This method tries to determine whether the flow should be restarted for certain statuses
+   * otherwise simply return.
+   * There are three scenarios that this method is called:
+   * 1. a flow is cleaned up by ContainerCleanupManager;
+   * 2. a flow has dispatch failure;
+   * 3. a flow encounters pod failure Each flow execution can be retried once.
+   * If the original status is EXECUTION_STOPPED or FAILED, then it will be retried only if
+   * "flow.retry.statuses" is defined with the status.
    *
    * @param flow
    * @param originalStatus
    */
-  public static void restartFlow(final ExecutableFlow flow, final Status originalStatus) {
+  static ExecutableFlow getFlowToRestart(final ExecutableFlow flow,
+      final Status originalStatus){
     final ExecutionOptions options = flow.getExecutionOptions();
-    // flow can only be retried once
-    if (options == null || options.isExecutionRetried()) {
-      return;
+    if (options == null) {
+      logger.warn("ExecutableFlow: " + flow.getExecutionId() + " has ExecutionOptions == null");
+      return null;
+    }
+    final Map<String, String> flowParams = options.getFlowParameters();
+    if (flowParams == null || flowParams.isEmpty()) {
+      logger.warn("ExecutableFlow: " + flow.getExecutionId() + " has ExecutionOptions == null");
+      return null;
     }
+
+    // flow can only retry if not reach the limit
+    final int flowMaxRetryLimit = Integer.parseInt(
+        flowParams.getOrDefault(FlowParameters.FLOW_PARAM_MAX_RETRIES, "0"));
+    if (flowMaxRetryLimit <= 0) {
+      logger.warn("ExecutableFlow: " + flow.getExecutionId() + " has exceed retry limit, count = "
+          + flowMaxRetryLimit);
+      return null;
+    }
+    // reduce the count
+    flowParams.put(FlowParameters.FLOW_PARAM_MAX_RETRIES, String.valueOf(flowMaxRetryLimit - 1));
```

Comment:
At this point we still don't know if there will be a retry or not (for example the method could return in line 219 without retrying). Why are we decreasing the counter already?